fast_llm/layers/language_model/head.py

```diff
@@ -405,31 +405,106 @@ def _logits_cross_entropy_forward_backward(
                 raise ValueError(
                     f"Invalid distillation loss implementation: {self._config.distillation_loss_implementation}"
                 )
-            distillation_loss = distillation_loss * self._config.distillation_loss_factor
         else:
             distillation_loss, distillation_grad = None, None
 
         # TODO: de-allocate earlier.
         del logits
+        loss, grad = self._post_process_loss_and_grad(
+            dpo_loss,
+            dpo_grad,
+            lm_loss,
+            lm_grad,
+            distillation_loss,
+            distillation_grad,
+            losses,
+            loss_mask,
+            kwargs,
+        )
 
-        # TODO: Accumulate grads in-place to reduce memory and compute overhead.
-        grad = _add_tensors(dpo_grad, lm_grad, distillation_grad)
+        return loss, output_parallel_linear_backward(grad, context) if self.training else None
 
-        # TODO: Return individual losses?
-        loss = _add_tensors(dpo_loss, lm_loss, distillation_loss)
-        if self.training and losses is not None:
-            if dpo_loss is not None:
+    def _post_process_loss_and_grad(
```

Comment:
This contains lots of unnecessary code. The gradients are conceptually ceparated from the losses, and processing them here adds unnecessary complexity to what is essentially call to `_add_tensors`.